### ISSUE TITLE
More generalize and more understandable `groupBy` method 

### ISSUE THREAD

**Issue by kanitw** [uwdata/compassql#122]
Current `groupBy` syntax in a CompassQL query looks like this: 

```js
{
  "spec": ...,
  "nest": [
    {"groupBy": "field"},
    {"groupBy": "fieldTransform"},
    {"groupBy": "encoding", "orderGroupBy": "effectiveness"},
    {"groupBy": "markAndTranspose", "orderGroupBy": "effectiveness"}
  ],
  "orderBy/chooseBy": "effectiveness",
  ...
}
```

The problem is that naming each `groupBy` key function to be information  can be quite confusing (e.g., `encoding`, `markAndTranspose` are pretty bad name). 

Instead of providing arbitrary name like above, I propose that we have `groupBy` as an array of property names with possibly transformation for further grouping (like how I group channel below) instead

```js
{
  "spec": ...,
  "nest": [
    /** replacing {"groupBy": "field"} */
    {"groupBy": [{property: "field"}]},

    /** replacing {"groupBy": "fieldTransform"} */
    {"groupBy": ["aggregate", "timeUnit", "bin"]},  
    // - "field" will be automatically included at this level  since we include it in the parent level

    /** replacing {"groupBy": "encoding", "orderGroupBy": "effectiveness"} */
    {
      "groupBy": [
        {
          property: "channel",  
          keys: {
             xy: ["x", "y"], 
             facet: ["row", "column"], 
             style: ["color", "size", "shape", "opacity"] 
          }
        }
      ],   
      // -  "field", "aggregate", "timeUnit", "bin" will be automatically included at this level since 
      //    we include them in the parent levels
      // - ":position, facet, style" suffix after "channel" means that 
      //   instead of using raw channel, we transform channel like
      //   x/y => "position", row/column => "facet", and  color/size/shape/opacity => "style"
      //   before using the transformed channel to group things.
      //   Therefore, x and y will be considered equivalent because 
      //   the "channel" value will become "position" instead of the original "x" or "y". 
      //   Similar for row/column => "facet" and color/size/shape/opacity => "style".
      "orderGroupBy": "effectiveness"
    },

    /** replacing {"groupBy": "markAndTranspose", "orderGroupBy": "effectiveness"} **/
    {
      "groupBy": [
         "mark",
         {property: "channel",  keys: { xy: ["x", "y"],  facet: ["row", "column"] }}
     ], 
      // - "field", "aggregate", "timeUnit", "bin" will be automatically included at this level 
      //   since we include them the parent levels  
      // - We specify channel with different transform, so it's overridden.
      //   Basically, channel transforms here doesn't include "style" 
      //   so "color/size/shape/opacity" are not grouped together at this level
    },
  ],
  "orderBy/chooseBy": "effectiveness",
  ...
}
```

This way, our groupBy framework can work with new properties (e.g., `scaleType`) that we are adding as well.  

@jheer, @domoritz, @ZeningQu  any opinions?

**Comment by kanitw** [comment]
With this framework,  we can automatically determine `groupBy`s for the Voyager2 UI with the following rules.  

0) Always have one level of grouping since our UI won't show hierarchy like the editor anyway. 

1) If there is a `?` field, group by `["field"]`. 
2) If there is no `?` field but there is a `?` fn (aggregate/timeUnit), group by `["aggregate", "timeUnit"]`

(Note: I don't think I will allow bin = `?` in the UI)

3) If there is no `?` field / fn but there is a `?` channel, group by `["channel: position, facet"]`

(If we decide that we will allow enumerating mark. we can add mark as level 4 - for now I don't think enumerating mark is that important for exploration)

**Comment by domoritz** [comment]
What’s a bit odd is that `{"groupBy": ["aggregate", "timeUnit", "bin"]},` is an array but `"groupBy": ["channel: position, facet, style"],` all channels are in one string. How is the group by different? I'd expect that it's written as `"groupBy": ["position", "facet", "style"],` or the same but with `channel:` prepended.

**Comment by kanitw** [comment]
"aggregate", "timeUnit", "bin", and "channel" are properties of a spec. 

"position", "facet", "style" are not a property of a spec; they are categories of a "channel".

**Comment by kanitw** [comment]
Implemented in #154.

We use `replace:  {x: 'xy', y:'xy', row:'facet': column: 'facet'}` instead of `keys` object described above.
